src/Datadog.Trace/TraceId.cs

```diff
@@ -0,0 +1,212 @@
+    public readonly struct TraceId : IEquatable<TraceId>
+    {
+        private readonly bool _is64Bit;
+        private readonly ulong _higher;
+        private readonly string _string;
+
+        private TraceId(ulong higher, ulong lower)
+        {
+            _is64Bit = false;
+            _higher = higher;
+            Lower = lower;
+            _string = $"{_higher:x16}{Lower:x16}";
+        }
+
+        private TraceId(ulong lower)
+        {
+            _is64Bit = true;
+            _higher = 0;
+            Lower = lower;
+            _string = Lower.ToString(CultureInfo.InvariantCulture);
+        }
+
+        /// <summary>
+        /// Gets TraceId with zero id.
+        /// </summary>
+        public static TraceId Zero => new(lower: 0);
+
+        /// <summary>
+        /// Gets lower 64 bits of 128 bit traceID or the whole 64 bit traceID.
+        /// </summary>
+        public ulong Lower { get; }
+
+        /// <summary>
+        /// Indicates if two specified instances of TraceId are not equal.
+        /// </summary>
+        /// <param name="traceId1">First <see cref="TraceId"/></param>
+        /// <param name="traceId2">Second <see cref="TraceId"/></param>
+        /// <returns>True if instances are equal, false otherwise.</returns>
+        public static bool operator ==(TraceId traceId1, TraceId traceId2)
+        {
+            return traceId1.Equals(traceId2);
+        }
+
+        /// <summary>
+        /// Indicates if two specified instances of TraceId are not equal.
+        /// </summary>
+        /// <param name="traceId1">First <see cref="TraceId"/></param>
+        /// <param name="traceId2">Second <see cref="TraceId"/></param>
+        /// <returns>True if instances are not equal, false otherwise.</returns>
+        public static bool operator !=(TraceId traceId1, TraceId traceId2)
+        {
+            return !(traceId1 == traceId2);
+        }
+
+        /// <summary>
+        /// Creates random 128 bit traceId.
+        /// </summary>
+        /// <returns>Instance of randomly generated <see cref="TraceId"/>.</returns>
+        public static TraceId CreateRandom()
+        {
+            var higherBytes = new byte[8];
+            var lowerBytes = new byte[8];
+
+            var randomNumberGenerator = RandomNumberGenerator.Current;
+            Unsafe.WriteUnaligned(ref higherBytes[0],  randomNumberGenerator.Next());
+            Unsafe.WriteUnaligned(ref lowerBytes[0], randomNumberGenerator.Next());
+
+            var higher = (ulong)BitConverter.ToInt64(higherBytes, startIndex: 0) & 0x7FFFFFFFFFFFFFFF;
+            var lower = (ulong)BitConverter.ToInt64(lowerBytes, startIndex: 0) & 0x7FFFFFFFFFFFFFFF;
+
+            return new TraceId(higher, lower);
+        }
+
+        /// <summary>
+        /// Creates random 64 bit traceId.
+        /// </summary>
+        /// <returns>Instance of randomly generated <see cref="TraceId"/>.</returns>
+        public static TraceId CreateRandom64Bit()
```

Comment:
Maybe this could also benefit from being named Datadog in some way. From this repo's perspective, this is only temporary and will be phased out, but it helps remove confusion because this method actually returns a trace id that only has 63-bits (the top bit is always zeroed out)